### PR TITLE
[Feature][WIP] Add option to download subtitles from Opensubtitles.org

### PR DESCRIPTION
**Pull request by Jacekun**
Do not merge yet. Just made this as a placeholder.

- Made it abstract to be future-proof and can be extended using any other subtitles API.

### CODE REVIEW

**Comment by Jacekun** [comment]
Mostly working.

What it Lacks:

- UI
  - **Username / Password / Language** setup on Settings page.
  - Button on player that shows new dialog where user can:
    - input query to search for, or type IMDB Id (auto filled when provider provides it) *(Calls ``search`` method)*.
    - pick which file to download. *(Populated by results from ``search`` method)*.
    - After picking a file, the dialog closes and returns to player with file downloaded and loaded into the video.

File: app/src/main/java/com/lagradost/cloudstream3/subtitles/AbstractSubProvider.kt [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import androidx.annotation.WorkerThread
+import com.lagradost.cloudstream3.subtitles.AbstractSubtitleEntities.*
+
+abstract class AbstractSubProvider{
+    open val name = ""
+
+    @WorkerThread
+    open suspend fun authorize(ouath: SubtitleOAuthEntity): SubtitleOAuthEntity {
+        throw NotImplementedError()
+    }
```
Comment (Blatzar) [inline app/src/main/java/com/lagradost/cloudstream3/subtitles/AbstractSubProvider.kt:12]: @LagradOst How about we add a `getAuthentication(context: Context)` function to generate the `SubtitleOAuthEntity` based on settings or the like.

**Comment by Jacekun** [comment]
UI development on https://github.com/Jacekun/CloudStream-3X/pull/2

**Comment by bachig26** [comment]
I know this project doesn’t make sense for cloudstream but can used for multi fetch, to look it out. https://github.com/a4k-openproject/a4kSubtitles

**Comment by Jacekun** [comment]
> I know this project doesn’t make sense for cloudstream but can used for multi fetch/ scraper, to look it out. _[ps. you may already know this]_ https://github.com/a4k-openproject/a4kSubtitles

Thanks! Will try to add Subscene as well using https://github.com/a4k-openproject/a4kSubtitles/blob/master/a4kSubtitles/services/subscene.py

**Comment by Blatzar** [comment]
@Jacekun This PR will likely take a while (a week) due to LagradOst being busy and he's got a lot of ideas about how to properly do the authentication system :skull: 
just a heads up

**Comment by LagradOst** [comment]
Yes, watch out for me :hando:

**Pull request merged**